libnavigation-base/src/main/java/com/mapbox/navigation/base/route/RouteAlternativesOptions.kt

```diff
@@ -0,0 +1,89 @@
+package com.mapbox.navigation.base.route
+
+import com.mapbox.api.directions.v5.models.RouteOptions
+import java.util.concurrent.TimeUnit
+
+/**
+ * The options available for observing alternative routes. The [RouteOptions] used to
+ * set the current active route are updated according to the current route progress.
+ *
+ * Register your [RouteAlternativesObserver] through [MapboxNavigation], the onRouteAlternatives
+ * callback will be triggered according every [intervalMillis] - while there is an active route.
+ *
+ * If there are no observers registered, alternatives will not be requested and these options
+ * are ignored.
+ *
+ * @param intervalMillis The interval in milliseconds, default is 5 min.
+ */
+class RouteAlternativesOptions private constructor(
+    val intervalMillis: Long
+) {
+    /**
+     * @return the builder that created the [RouteAlternativesOptions]
+     */
+    fun toBuilder(): Builder = Builder().apply {
+        intervalMillis(intervalMillis)
+    }
+
+    /**
+     * Regenerate whenever a change is made
+     */
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as RouteAlternativesOptions
+
+        if (intervalMillis != other.intervalMillis) return false
+
+        return true
+    }
+
+    /**
+     * Regenerate whenever a change is made
+     */
+    override fun hashCode(): Int {
+        return intervalMillis.hashCode()
+    }
+
+    /**
+     * Returns a string representation of the object.
+     */
+    override fun toString(): String {
+        return "RouteRefreshOptions(" +
+            "intervalMillis=$intervalMillis" +
+            ")"
+    }
+
+    /**
+     * Build your [RouteAlternativesOptions].
+     */
+    class Builder {
+        private var intervalMillis: Long = TimeUnit.MINUTES.toMillis(5)
+
+        /**
+         * Update the route refresh interval in milliseconds.
+         */
+        fun intervalMillis(intervalMillis: Long): Builder {
+            this.intervalMillis = intervalMillis
+            return this
+        }
+
+        /**
+         * Build the object.
+         */
+        fun build(): RouteAlternativesOptions {
+            check(intervalMillis >= MINIMUM_REFRESH_INTERVAL) {
+                "Route alternatives interval out of range" +
+                    " $intervalMillis < $MINIMUM_REFRESH_INTERVAL"
+            }
+            return RouteAlternativesOptions(
+                intervalMillis = intervalMillis
+            )
+        }
+    }
+
+    private companion object {
+        private val MINIMUM_REFRESH_INTERVAL = TimeUnit.SECONDS.toMillis(10)
```

Comment:
Isn't that too often? I'm worried about the overhead that developers could even unintentionally cause.

Reply:
I agree. What about battery power consumption?

Reply:
Our default value is 5 minutes. The minimum allowed is 10 seconds. Customer feedback is typically, asking for lower lower thresholds.

For what it's worth, Lyft requested directions from Google once every 5 seconds in their driver app. Battery was not a concern because drivers are always plugged in.

Reply:
I'm thinking more in terms of not flooding Directions API.

@kshehadeh @danpat we're reworking here a feature that periodically requests alternative routes (by making a new route request). What would be your recommendation for a minimal interval between those requests?

Reply:
@LukasPaczos Rather than hard-coding a refresh interval, I would do the following:

  - Examine the `Cache-Control` header and look at the `max-age` value.  Set the refresh timer to trigger when `max-age` runs out.
  - If `max-age` is not set, look for an `Expires` header - if set, set the fresh timer to trigger when `Expires` occurs.
  - If neither header is set, use 60 seconds.

We do set `Cache-Control` on all API responses currently, and intend to do so indefinitely.  Using these headers will allow us to optimize the experience for all customers without triggering excessive refresh requests at a resolution greater than our traffic/incident data is updating.

Reply:
This is different from a refresh interval. This is, examining the alternatives while in-route. So the origin location is changing.